lib/HttpClient.ts

```diff
@@ -186,48 +200,68 @@ export class HttpClient implements ifm.IHttpClient {
         }
 
         let info: RequestInfo = this._prepareRequest(verb, requestUrl, headers);
-        let response: HttpClientResponse = await this.requestRaw(info, data);
 
-        // Check if it's an authentication challenge
-        if (response && response.message && response.message.statusCode === HttpCodes.Unauthorized) {
-            let authenticationHandler: ifm.IRequestHandler;
+        // Only perform retries on reads since writes may not be idempotent.
+        let maxTries: number = (this._allowRetries && RetryableHttpVerbs.indexOf(verb) != -1) ? this._maxRetries + 1 : 1;
```

Comment:
I think its more clear to put the retryable http verbs part in the while loop. Otherwise you have to read that then look back up here.

Reply:
We need the contents of the loop to run at least once regardless of the verb though. So if we move it into the while condition we'd have to do some weird special casing that would make it more confusing IMO.